README.md

```diff
@@ -20,9 +20,9 @@ Log into that VM using the "ubuntu" user, and run these commands:
     $ git clone https://github.com/delphix/appliance-build.git
     $ cd appliance-build
     $ ansible-playbook bootstrap/playbook.yml
-    $ ./scripts/docker-run.sh make internal-minimal
+    $ HYPERVISORS=generic ./scripts/docker-run.sh make internal-minimal
```

Comment:
How much thought have we given to the term of "hypervisors" here? Personally, it seems a little odd since we're not really listing hypervisors; e.g. "azure" isn't really the hypervisor name vs. "hyperv".

We also have the same concept in the linux-pkg repository, and I think the term "platform" is used there.

It seems like ubuntu uses the term "kernel flavour" for this; see [here](https://wiki.ubuntu.com/Kernel/Dev/Flavours).. which is a little confusing, since it also has the term "ubuntu flavour" which is different; see [here](https://www.ubuntu.com/download/flavours)

I think it'd be good to collectively decide on a good term to use here, and be consistent across our various projects/repositories. While it might not seem important, correct/consistent terminology can make future discussion more clear, since the term will be well defined across teams/projects/repos.

Personally, I lean towards something like "flavour" since it's already used by ubuntu (although we'd be overloading it even more), and it doesn't come with any baggage inside delphix (nothing inside delphix is coined "flavour" AFAIK, vs. hypervisor, where folks might have a preconceived notion of what that means); "platform" might also be good, but that's already used in the "delphix dynamic data platform".

Using something more ambiguous like "flavour" vs "hypervisor" seems more appropriate when we'll likely create different images for things that'll use the same underlying hypervisor; e.g. google cloud which is kvm based, vs. other kvm based cloud providers (openstack?).

Thoughts?

Reply:
It's probably also worth noting that the app-gate's `server.conf` has a `platform` field, which I think is used to denote this type of thing; e.g. see [here](https://gitlab.delphix.com/app/dlpx-app-gate/blob/master/appliance/server/sh/boot/dxinit.sh#L334). So, perhaps "platform" would actually be a good choice of terminology since it's already used.

Reply:
I agree that we should try to come up with a common terminology across all the projects. In `linux-pkg` we are using `kernel` for a specific kernel version and `platform` for a kernel "flavour". I think there's a distinction between the "kernel flavour" and what we refer to as the "hypervisor" here, so we should probably not treat them as the same. We could have 2 different "hypervisors" use the same "kernel flavour", e.g. let's say we could have 2 different builds for `OpenStack` and `CloudStack`, which both use the `kvm` kernel.

As for the naming, I think we'd need to discuss further.

Reply:
I thought that I replied to this before, but I must have never published my comment.

I'm definitely open to using another term. My feeling is that it is actually better to stay away from something vague like 'flavour', since that basically means the same thing as 'variant' so it isn't obvious how the two relate to each other. We could use something like different types of 'variant', e.g. 'product variant' and 'platform variant'. I'm not sure if that would do a better job at conveying what these two concepts are and that they orthogonal to each other.

I also like 'platform', since I think that does a good job of capturing the concept introduced here. I don't worry too much about confusion with 'delphix dynamic data platform', since I don't think that term is likely to crop up too often in this context. @pzakha, I think both `linux-pkg` and `appliance-build` face the issue you point out. If we ever do have different builds which use the same 'kernel flavor', we could need to make that distinction in `delphix-platform` if the two different builds include different packages. If we do want to use 'platform', we could just punt on that issue for now in both repositories.

Reply:
Okay, so let's agree on the terms we want to use for different things. I propose:
Product variant: `variant`, e.g. internal-dev, internal-qa
Cloud variant: `platform`, e.g. ESX, AWS, CloudStack
Kernel flavour: `kernel type`, e.g. generic, aws, kvm
Specific kernel version: `kernel version`, e.g. 4.15.0-18-generic

So we'd keep using `variant` for the product variant, we'd change `platform` to `kernel type` in `linux-pkg`, and we'd change `kernel` to `kernel version` in linux-pkg. Finally, we'd use `platform` for the target cloud platform (or hypervisor in the case of ESX). delphix-platform will be customizable for each `platform`.

We currently have `KVERS` and `PLATFORM` placeholders in a few packages. We can keep `KVERS` and change `PLATFORM` to `KTYPE`.

For the `kernel type`, we could also use `kernel flavour` or simply `kernel`.

Reply:
> Product variant: variant, e.g. internal-dev, internal-qa

This is awfully similar to "product types", and likely easy to mix up; e.g. standard, aws marketplace, etc.

Reply:
@prakashsurya So are you proposing we change the current term `variant` to refer to things like internal-dev?
Just in case it wasn't clear, the terms I'm proposing to use are:
`variant`, `platform`, `kernel type`, `kernel version`

Reply:
I'm commenting on "product variant", and how that sounds similar to "product type". I wouldn't be surprised if folks confuse the two due to their similarities (both in their name, but also in how the two are implemented in the build), that's all.

It'd be good to loop in the notion of our product type in this list, but that's kind of encoded in variant (at least right now, this may not be the case moving forward).

> So are you proposing we change the current term variant to refer to things like internal-dev?

I don't understand; isn't this how you defined it above? "product variant".. "e.g. internal-dev internal-qa".. I coined the term variant in the appliance-build repository to mean these different build variations, so I'm not implying we change that. Just that this term is distinct from our product types, and they warrant terms distinct enough to avoid accidental confusion.

We could have a "dev" variant of our "standard" product type (what internal-dev is), but we could also have a "dev" variant of our "aws marketplace" product type (which we don't currently build). We could also have an "external" variant of our "standard" product type (what external-standard is). Basically, as it stands now, we could potentially want to produce the cross product of build variants (which is "external", "dev", and "qa") and product types (which is "standard", "aws marketplace", etc).

My suggestion would be to use the terms "build variant" and "product type".

Reply:
Okay, so just to clarify, as it looks like I've caused some confusion, here are the terms I'm proposing:
`variant`: e.g. internal-dev, internal-qa
`platform`: e.g. ESX, AWS, CloudStack
`kernel type`: e.g. generic, aws, kvm
`kernel version`: e.g. 4.15.0-18-generic

If we want to subdivide `variant` into `build variant` and `product type`, I'm fine with that. I've mostly been focusing on the other 3 terms.

Reply:
This terminology sounds good to me.

So in appliance-build, we would expose `platform` as a parameter that can be passed in, and then `kernel type` would be chosen automatically based on the `platform`, and `kernel version` would be chosen implicitly by specifying the version of `linux-pkg` to use.

How do we see this working in `linux-pkg` though? It seems like we would need to make `platform` a parameter. Would `kernel version` need to be configurable as well?

Reply:
The way I see it, `appliance-build` doesn't need to care at all about the kernel, only about the platform. `delphix-platform` is the one that will install the proper kernel via dependencies.

For `linux-pkg`, it doesn't need to know anything about platforms. It only cares about `kernel version`s and `kernel type`s. When linux-pkg builds delphix-platform, delphix-platform will be the one that is aware of all the platforms and will produce one deb for each.

Reply:
`appliance-build` does need to care about to the kernel, because you have to specify the `kernel-flavour` when running live-build, otherwise you end up with the generic kernel in addition to the one specified by `delphix-platform`. That's pretty easy to handle though. We can either hardcode the mapping of `platform` to `kernel type`, or just get the `kernel type` out of the `delphix-platform` package's metadata.

I'm still not sure what `linux-pkg`'s behavior will be though. If someone runs `linux-pkg` with just `gcp` specified for `kernel type`, will `delphix-platform` decide it should only produce a `delphix-platform-google` package?

Reply:
> appliance-build does need to care about to the kernel, because you have to specify the kernel-flavour when running live-build, otherwise you end up with the generic kernel in addition to the one specified by delphix-platform
I'm not sure to understand, does appliance-build try to install a kernel manually somewhere? The idea is that appliance-build doesn't install the kernel itseld but rather relies on `delphix-platform` to pull it in.

> If someone runs linux-pkg with just gcp specified for kernel type, will delphix-platform decide it should only produce a delphix-platform-google package?
The `kernel type` field is mostly there so that we can override the default kernel versions. For normal runs of `linux-pkg` (i.e. those that will be consumed by appliance-build) it *has* to build packages for all supported kernel types.

Reply:
> I'm not sure to understand, does appliance-build try to install a kernel manually somewhere?

Yeah, exactly: https://github.com/delphix/appliance-build/pull/166/files#diff-c6618c419db1b442c4c804d5cf7b36abR104. I don't think this is an issue though, `delphix-platform` can still be the source of truth for the `kernel type`.

> The kernel type field is mostly there so that we can override the default kernel versions.

So `linux-pkg` will need to build for every kernel type on every build, and when run manually, some of the kernel types may have their versions overridden? That seems reasonable to me.

Reply:
Nice find! That doesn't seem to be documented in the man page. I'll try it out.

Reply:
Yeah, I'd recommend getting the source package for live-build on Ubuntu and then extracting it on your laptop. It helped me a lot when looking for some customizations of the build since the man pages are far from being comprehensive.